src/main/java/org/jenkinsci/plugins/workflow/support/storage/SimpleXStreamFlowNodeStorage.java

```diff
@@ -175,17 +168,14 @@ private void storeActions() {  // We've already loaded the actions, may as well
             @Override public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext context) {
                 try {
                     FlowNode n = (FlowNode) ref.unmarshal(reader, context);
-                    ids.put(n, reader.getValue());
                     try {
                         @SuppressWarnings("unchecked") List<FlowNode> parents = (List<FlowNode>) FlowNode$parents.get(n);
                         if (parents != null) {
                             @SuppressWarnings("unchecked") List<String> parentIds = (List<String>) FlowNode$parentIds.get(n);
                             assert parentIds == null;
                             parentIds = new ArrayList<String>(parents.size());
                             for (FlowNode parent : parents) {
-                                String id = ids.get(parent);
-                                assert id != null;
-                                parentIds.add(id);
+                                parentIds.add(parent.getId().intern());
```

Comment:
The problem here is that when we are reading an old build record, `parent.getId() == null`.